### PR TITLE
parser,ast: parse `admin show slow` statement

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

Add the following new syntax:
* ```admin show slow top [internal | all] N```
* ```admin show slow recent N```

I expect to use it like this:

```
mysql> admin show log top 3;
+-------------------------------------+----------------------------+----------+------------------------------------------------------------------------------------------------------+------+---------+--------------------+----------------+------+-----------------+-----------+----------+
| SQL                                 | START                      | DURATION | DETAILS                                                                                              | SUCC | CONN_ID | TXNTS              | USER           | DB   | TABLE_IDS       | INDEX_IDS | INTERNAL |
+-------------------------------------+----------------------------+----------+------------------------------------------------------------------------------------------------------+------+---------+--------------------+----------------+------+-----------------+-----------+----------+
| select sleep(2)                     | 2018-09-13 14:18:39.723710 | 00:00:02 |                                                                                                      |    1 |       1 | 402868016168042497 | root@127.0.0.1 | test |                 |           |        1 |
| select count(*) from t limit 330000 | 2018-09-13 14:20:38.970281 | 00:00:08 | process_time:1m0.665s wait_time:28.216s request_count:14 total_keys:20000014 processed_keys:20000000 |    1 |       1 | 402868047428714497 | root@127.0.0.1 | test | table_ids:[32], |           |        1 |
| select count(*) from t              | 2018-09-13 14:17:45.066304 | 00:00:09 | process_time:1m5.771s wait_time:31.32s request_count:14 total_keys:20000014 processed_keys:20000000  |    1 |       1 | 402868001842397185 | root@127.0.0.1 | test | table_ids:[32], |           |        1 |
+-------------------------------------+----------------------------+----------+------------------------------------------------------------------------------------------------------+------+---------+--------------------+----------------+------+-----------------+-----------+----------+
3 rows in set (0.00 sec)
```

```
mysql> admin show log recent 3;
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------+----------+------------------------------------------------------------------------------------------------------+------+---------+--------------------+----------------+------+-----------------+----------------+----------+
| SQL                                                                                                                                                                               | START                      | DURATION | DETAILS                                                                                              | SUCC | CONN_ID | TXNTS              | USER           | DB   | TABLE_IDS       | INDEX_IDS      | INTERNAL |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------+----------+------------------------------------------------------------------------------------------------------+------+---------+--------------------+----------------+------+-----------------+----------------+----------+
| select count(*) from t limit 110000                                                                                                                                               | 2018-09-13 14:20:55.715194 | 00:00:09 | process_time:1m6.188s wait_time:30.196s request_count:14 total_keys:20000014 processed_keys:20000000 |    1 |       1 | 402868051819626497 | root@127.0.0.1 | test | table_ids:[32], |                |        1 |
| SELECT version, table_id, modify_count, count from mysql.stats_meta where version > 400040410672791554 order by version                                                           | 2018-09-13 14:20:57.148856 | 00:00:03 | wait_time:3.884s request_count:2 total_keys:266 processed_keys:144                                   |    1 |       0 | 402868052199735298 |                |      | table_ids:[19], | index_ids:[1], |        0 |
| select variable_name, variable_value from mysql.global_variables where variable_name in ('tidb_auto_analyze_ratio', 'tidb_auto_analyze_start_time', 'tidb_auto_analyze_end_time') | 2018-09-13 14:20:57.148658 | 00:00:03 | wait_time:3.884s request_count:1 total_keys:3                                                        |    1 |       0 | 402868052199735297 |                |      | table_ids:[13], | index_ids:[1], |        0 |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------+----------+------------------------------------------------------------------------------------------------------+------+---------+--------------------+----------------+------+-----------------+----------------+----------+
3 rows in set (0.00 sec)
```

### What is changed and how it works?

Update `parser.y`

### Check List 

 - Unit test

Related changes

 - Need to update the documentation

PTAL @winkyao @coocood

### CODE REVIEW

File: parser/parser.y [hunk 1]
```diff
@@ -3606,6 +3611,10 @@ FunctionCallNonKeyword:
 			Args: []ast.ExprNode{ast.NewValueExpr($3), $5},
 		}
 	}
+|	"LOG" '(' ExpressionListOpt ')'
```
Comment (coocood) [inline parser/parser.y]: Why add this?

File: parser/parser.y [hunk 2]
```diff
@@ -5221,6 +5230,53 @@ AdminStmt:
 			JobIDs: $6.([]int64),
 		}
 	}
+|	"ADMIN" "SHOW" "LOG" AdminShowLog
```
Comment (coocood) [inline parser/parser.y]: "LOG" is too general, "SLOW_QUERY" is better.

File: parser/parser.y (hunk 1 above)
Comment (tiancaiamao) [reply parser/parser.y]: After this PR, "LOG" change to a token from an identifier, it's a not keyword token.
There is no rule for builtin function "select log(...)" , so I add this rule.

File: parser/parser.y (hunk 2 above)
Comment (tiancaiamao) [reply parser/parser.y]: @winkyao What's your opinion ?

File: parser/parser.y [hunk 3]
```diff
@@ -5221,6 +5230,53 @@ AdminStmt:
...
+	{
+		$$ = &ast.AdminStmt{
+			Tp: ast.AdminShowLog,
+			ShowLog: $4.(*ast.ShowLog),
+		}
+	}
+
+AdminShowLog:
+	"RECENT" NUM
```
Comment (coocood) [inline parser/parser.y]: We can remove the `RECENT` and make the `NUM` optional

Comment (tiancaiamao) [reply parser/parser.y]: There will be two containers, one heap for TopN , one queue for Recent.
Suggested by @shenli

File: parser/parser.y [hunk 4]
```diff
@@ -5221,6 +5230,53 @@ AdminStmt:
...
+		}
+	}
+|	"TOP" NUM
+	{
+		$$ = &ast.ShowLog{
+			Tp: ast.ShowLogTop,
+			Count: getUint64FromNUM($2),
+		}
+	}
+|	"TOP" "USER" NUM
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply parser/parser.y]: Good point!

File: parser/parser.y [hunk 5]
```diff
@@ -5221,6 +5230,53 @@ AdminStmt:
...
+	{
+		$$ = &ast.ShowLog{
+			Tp: ast.ShowLogTop,
+			Kind: "user",
+			Count: getUint64FromNUM($3),
+		}
+	}
+|	"TOP" "INTERNAL" NUM
```
Comment (coocood) [inline parser/parser.y]: This syntax can be removed, use `ALL` is enough.

**Comment by shenli** [comment]
`Admin show slow query ....` is more readable.

**Comment by shenli** [comment]
The value in `TABLE_IDS ` and `INDEX_IDS` do not have to contain `table_ids:` and `index_ids:` as prefixes.
```sql
mysql> admin show log recent 3;
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------+----------+------------------------------------------------------------------------------------------------------+------+---------+--------------------+----------------+------+-----------------+----------------+----------+
| SQL                                                                                                                                                                               | START                      | DURATION | DETAILS                                                                                              | SUCC | CONN_ID | TXNTS              | USER           | DB   | TABLE_IDS       | INDEX_IDS      | INTERNAL |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------+----------+------------------------------------------------------------------------------------------------------+------+---------+--------------------+----------------+------+-----------------+----------------+----------+
| select count(*) from t limit 110000                                                                                                                                               | 2018-09-13 14:20:55.715194 | 00:00:09 | process_time:1m6.188s wait_time:30.196s request_count:14 total_keys:20000014 processed_keys:20000000 |    1 |       1 | 402868051819626497 | root@127.0.0.1 | test | table_ids:[32], |                |        1 |
| SELECT version, table_id, modify_count, count from mysql.stats_meta where version > 400040410672791554 order by version                                                           | 2018-09-13 14:20:57.148856 | 00:00:03 | wait_time:3.884s request_count:2 total_keys:266 processed_keys:144                                   |    1 |       0 | 402868052199735298 |                |      | table_ids:[19], | index_ids:[1], |        0 |
| select variable_name, variable_value from mysql.global_variables where variable_name in ('tidb_auto_analyze_ratio', 'tidb_auto_analyze_start_time', 'tidb_auto_analyze_end_time') | 2018-09-13 14:20:57.148658 | 00:00:03 | wait_time:3.884s request_count:1 total_keys:3                                                        |    1 |       0 | 402868052199735297 |                |      | table_ids:[13], | index_ids:[1], |        0 |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------+----------+------------------------------------------------------------------------------------------------------+------+---------+--------------------+----------------+------+-----------------+----------------+----------+
3 rows in set (0.00 sec)
```

File: parser/parser.y (hunk 3 above)
Comment (coocood) [reply parser/parser.y]: No "TOP" means recent.

Comment (shenli) [reply parser/parser.y]: TOP doesn't mean `recent`. It means `order by x limit n`.

File: parser/parser.y (hunk 2 above)
Comment (shenli) [reply parser/parser.y]: I vote for the `slow query`.

File: parser/parser.y (hunk 5 above)
Comment (coocood) [reply parser/parser.y]: Or Keep this and remove "ALL"

File: parser/parser.y (hunk 2 above)
Comment (ciscoxll) [reply parser/parser.y]: Support for using  `"LOG"` .

File: parser/parser.y (hunk 5 above)
Comment (tiancaiamao) [reply parser/parser.y]: How to represent the result "user" + "internal" ?
We have three cases:
* "user"
* "internal"
* "user" + "internal"

 one keyword is not enough

File: parser/parser.y (hunk 2 above)
Comment (morgo) [reply parser/parser.y]: How about `ADMIN SHOW SLOW`?  I agree that log is too generic, but with `slow` it could be implied.

**Comment by tiancaiamao** [comment]
PTAL @coocood @shenli

File: ast/misc.go [hunk 6]
```diff
@@ -626,6 +627,26 @@ type HandleRange struct {
...
+	ShowSlowRecent
+)
+
+// ShowSlow is used for the following command:
+//	admin show slow top [ internal | all] N
+//	admin show slow recent N
+type ShowSlow struct {
+	Tp    ShowSlowType
+	Count uint64
+	// May be "internal" or "all", default is ""
```
Comment (winkyao) [inline ast/misc.go]: what about only showing general type?

Comment (tiancaiamao) [reply ast/misc.go]: What do you mean by "general" type? is that "user + internal"?
"user" == "" (the default)
"user + internal" == "all"

**Comment by coocood** [comment]
LGTM

File: parser/parser.y [hunk 7]
```diff
@@ -5221,6 +5226,45 @@ AdminStmt:
...
+		$$ = &ast.ShowSlow{
+			Tp: ast.ShowSlowTop,
+			Count: getUint64FromNUM($2),
+		}
+	}
+|	"TOP" "INTERNAL" NUM
+	{
+		$$ = &ast.ShowSlow{
+			Tp: ast.ShowSlowTop,
+			Kind: "internal",
```
Comment (crazycs520) [inline parser/parser.y]: How about define a `const` :  KindInternal, KindAll, KindDefault ?

File: parser/parser.y [hunk 8]
```diff
@@ -5221,6 +5226,45 @@ AdminStmt:
...
+			Count: getUint64FromNUM($3),
+		}
+	}
+|	"TOP" "ALL" NUM
+	{
+		$$ = &ast.ShowSlow{
+			Tp: ast.ShowSlowTop,
+			Kind: "all",
```
Comment (crazycs520) [inline parser/parser.y]: dito

File: ast/misc.go (hunk 6 above)
Comment (winkyao) [reply ast/misc.go]: Please add some comment to explain what  `"user"` is, and the default.

Comment (tiancaiamao) [reply ast/misc.go]: I've defined a `ShowSlowKind` type, instead of using string. @winkyao @crazycs520

**Pull request merged**